File: src/main/java/com/campspot/lib/PunchLib.java
package com.campspot.lib;

import com.campspot.api.Punch;
import com.campspot.dao.PunchDAO;
import org.joda.time.DateTime;

import java.util.List;

public class PunchLib {
  private final PunchDAO punchDAO;

  public PunchLib(PunchDAO punchDAO) {
    this.punchDAO = punchDAO;
  }

  public Punch create(Punch punch) {
    validatePunchDoesNotOverlap(punch);

    Long id = punchDAO.create(punch.getStart(), punch.getEnd(), punch.getCategory(), punch.getDescription());
    return punchDAO.findById(id);
  }

  public Punch update(Punch punch) {
    validatePunchDoesNotOverlap(punch);

    punchDAO.update(punch.getId(), punch.getStart(), punch.getEnd(), punch.getCategory(), punch.getDescription());

    return punchDAO.findById(punch.getId());
  }

  public List<Punch> listForDates(DateTime start, DateTime end) {
    return punchDAO.listForDates(start, end);
  }

  private void validatePunchDoesNotOverlap(Punch punch) {
    if (punchDAO.anyInRange(punch.getStart(), punch.getEnd())) {
      throw new RuntimeException("Punch Cannot Overlap");
    }
  }
}

File: src/main/java/com/campspot/lib/PunchLib.kt
package com.campspot.lib

import com.campspot.api.Punch
import com.campspot.dao.PunchDAO
import org.joda.time.DateTime

class PunchLib(private val punchDAO: PunchDAO) {
  fun create(punch: Punch): Punch {
    validatePunchDoesNotOverlap(punch)

    val id = punchDAO.create(punch.start, punch.end, punch.category, punch.description)
    return punchDAO.findById(id)
  }

  fun update(punch: Punch): Punch {
    validatePunchDoesNotOverlap(punch)

    punchDAO.update(punch.id, punch.start, punch.end, punch.category, punch.description)

    return punchDAO.findById(punch.id)
  }

  fun listForDates(start: DateTime, end: DateTime): List<Punch> {
    return punchDAO.listForDates(start, end)
  }

  private fun validatePunchDoesNotOverlap(punch: Punch) {
    if (punchDAO.anyInRange(punch.start, punch.end)!!) {
      throw RuntimeException("Punch Cannot Overlap")
    }
  }
}
